server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/RemoteShardsBalancer.java

```diff
@@ -0,0 +1,598 @@
+    }
+
+    public Map<String, UnassignedIndexShards> groupUnassignedShardsByIndex() {
+        HashMap<String, UnassignedIndexShards> unassignedShardMap = new HashMap<>();
+        unassignIgnoredRemoteShards(allocation);
+        for (ShardRouting shard : routingNodes.unassigned().drain()) {
+            String index = shard.getIndexName();
+            if (!RoutingPool.REMOTE_CAPABLE.equals(RoutingPool.getShardPool(shard, allocation))) {
+                routingNodes.unassigned().add(shard);
+                continue;
+            }
+            if (!unassignedShardMap.containsKey(index)) {
+                unassignedShardMap.put(index, new UnassignedIndexShards());
+            }
+            unassignedShardMap.get(index).addShard(shard);
+        }
+        return unassignedShardMap;
+    }
+
+    /**
+     * Unassigned shards from {@link LocalShardsBalancer} are ignored since the balancer cannot allocate remote shards.
+     * Prior to allocation operations done by {@link RemoteShardsBalancer}, the ignored remote shards are moved back to
+     * unassigned status.
+     */
+    private void unassignIgnoredRemoteShards(RoutingAllocation routingAllocation) {
+        RoutingNodes.UnassignedShards unassignedShards = routingAllocation.routingNodes().unassigned();
+        for (ShardRouting shard : unassignedShards.drainIgnored()) {
+            RoutingPool pool = RoutingPool.getShardPool(shard, routingAllocation);
+            if (pool == RoutingPool.REMOTE_CAPABLE && shard.unassigned() && (shard.primary() || !shard.unassignedInfo().isDelayed())) {
+                unassignedShards.add(shard);
+            } else {
+                unassignedShards.ignoreShard(shard, shard.unassignedInfo().getLastAllocationStatus(), routingAllocation.changes());
+            }
+        }
+    }
+
+    private void allocateUnassignedPrimaries(Queue<RoutingNode> nodeQueue, Map<String, UnassignedIndexShards> unassignedShardMap) {
+        allocateUnassigned(true, nodeQueue, unassignedShardMap);
+    }
+
+    private void allocateUnassignedReplicas(Queue<RoutingNode> nodeQueue, Map<String, UnassignedIndexShards> unassignedShardMap) {
+        allocateUnassigned(false, nodeQueue, unassignedShardMap);
+    }
+
+    private void ignoreRemainingShards(Map<String, UnassignedIndexShards> unassignedShardMap) {
+        for (UnassignedIndexShards indexShards : unassignedShardMap.values()) {
+            for (ShardRouting shard : indexShards.getPrimaries()) {
+                routingNodes.unassigned().ignoreShard(shard, UnassignedInfo.AllocationStatus.DECIDERS_NO, allocation.changes());
+            }
+            for (ShardRouting shard : indexShards.getReplicas()) {
+                routingNodes.unassigned().ignoreShard(shard, UnassignedInfo.AllocationStatus.DECIDERS_NO, allocation.changes());
+            }
+        }
+    }
+
+    private void allocateUnassigned(
```

Comment:
Can we break this down to smaller methods? Looks difficult to follow.

Reply:
Done.